chore(*): switch from JSHint/JSCS to ESLint (+ other build-related changes)

mgol opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
A build-related refactor.

**What is the current behavior? (You can also link to an open issue here)**
Code is linted via JSHint+JSCS.

**What is the new behavior (if this is a feature change)?**
Code is linted via ESLint.

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] ~~Tests for the changes have been added (for bug fixes / features)~~
- [x] ~~Docs have been added / updated (for bug fixes / features)~~

**Other information**:
This is not ready (there are still many ESLint violations) but since the changes are rather large, I'd like to get some feedback now.

There are a couple of build-related commits; submitting them as separate PRs would be difficult as they modify `package.json` and shrinkwrap file. It makes sense to land them separately and all but the last commit are small so it shouldn't case lots of problems in reviewing.

mgol committed:
chore(package.json): add grunt-cli to devDependencies, update grunt-* packages

The Travis build.sh script has to be updated as the previous way of
specifying parameters to the code run in Grunt tasks stopped working
with the newest Grunt.

mgol committed:
chore(*): Add .nvmrc

If one uses nvm to manage Node.js versions, the .nvmrc file makes `nvm use`
switch to the version specified in .nvmrc. There are scripts that invoke
it automatically when cd'ing to directories containing .nvmrc so that you
never run build commands using a wrong Node version, see:
https://github.com/creationix/nvm/blob/v0.31.2/README.markdown#zsh

mgol committed:
chore(package.json): remove the engineStrict field

The engineStrict field is deprecated in npm 2 and removed in npm 3.

mgol committed:
chore(*): switch from JSHint/JSCS to ESLint

Thanks to @narretz for help in fixing style violations and to @gkalpak
for a very extensive review.

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+    "accessor-pairs": "error",
+    "array-callback-return": "error",
+    "complexity": ["warn", 10],
+//    "dot-notation": "error", // TODO enable?
```
I think there are places where we deliberately use bracket-notation. If there are many such places (that need a "manual" override), I would leave this disabled.

mgol replied:
I was thinking about moving all those `TODO` rules to a separate
.eslintrc-todo.json file, then discuss all the options, remove the one we
don't want to enable and the rest will stay there, waiting for someone to
fix the code and enable the rule.

Thoughts?

On Tue, Jul 26, 2016 at 1:32 PM, Georgios Kalpakas <notifications@github.com
> wrote:

> In .eslintrc.json
> <https://github.com/angular/angular.js/pull/14952#discussion_r72231300>:
>
> > +    "no-irregular-whitespace": "error",
> > +    "no-negated-in-lhs": "error",
> > +    "no-obj-calls": "error",
> > +    "no-regex-spaces": "error",
> > +    "no-sparse-arrays": "error",
> > +    "no-unreachable": "error",
> > +    "use-isnan": "error",
> > +    "no-unsafe-finally": "error",
> > +    "valid-typeof": "error",
> > +    "no-unexpected-multiline": "error",
> > +
> > +    // Best practices
> > +    "accessor-pairs": "error",
> > +    "array-callback-return": "error",
> > +    "complexity": ["warn", 10],
> > +//    "dot-notation": "error", // TODO enable?
>
> I think there are places where we deliberately use bracket-notation. If
> there are many such places (that need a "manual" override), I would leave
> this disabled.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/6f56b890138f16a67b5f61cb2238d75378a76497..22f65cbbbf36f1ae396038f457ca2266f5b52767#r72231300>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnvg_rt1IyxEc7jYUKkY9aFjn3a20ks5qZfBkgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak replied:
SGTM

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+    "array-callback-return": "error",
+    "complexity": ["warn", 10],
+//    "dot-notation": "error", // TODO enable?
+//    "dot-location": ["error", "property"], // TODO enable?
```
I think we have several occurences of both. I personally prefer having the dot on the same line as the object.
I think we can leave this disabled for start and consider it as a follow up.

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+
+    // Stylistic issues
+    "array-bracket-spacing": ["error", "never"],
+//    "block-spacing": ["error", "always"], // TODO enable?
```
I like the idea of enabling this, but am fine leaving it for later :smiley:

Narretz replied:
Is this the rule for spaces at the beginning and the end of a block?

mgol replied:
>
> > +// "block-spacing": ["error", "always"], // TODO enable?
>
> Is this the rule for spaces at the beginning and the end of a block?
>
Yup. http://eslint.org/docs/rules/block-spacing

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+    "array-bracket-spacing": ["error", "never"],
+//    "block-spacing": ["error", "always"], // TODO enable?
+    "brace-style": ["error", "1tbs", { "allowSingleLine": true }],
+//    "comma-spacing": "error", // TODO enable?
```
Same as above.

mgol replied:
"Above" is hard to understand when you're reading it in email
notifications. :D

On Tue, Jul 26, 2016 at 1:38 PM, Georgios Kalpakas <notifications@github.com
> wrote:

> In .eslintrc.json
> <https://github.com/angular/angular.js/pull/14952#discussion_r72232076>:
>
> > +    "no-delete-var": "error",
> > +    "no-label-var": "error",
> > +    "no-restricted-globals": ["error", "event"],
> > +    "no-shadow-restricted-names": "error",
> > +    "no-undef-init": "error",
> > +    "no-undef": "error",
> > +    "no-unused-vars": ["error", { "vars": "local", "args": "none" }],
> > +
> > +    // Node.js
> > +    "handle-callback-err": "error",
> > +
> > +    // Stylistic issues
> > +    "array-bracket-spacing": ["error", "never"],
> > +//    "block-spacing": ["error", "always"], // TODO enable?
> > +    "brace-style": ["error", "1tbs", { "allowSingleLine": true }],
> > +//    "comma-spacing": "error", // TODO enable?
>
> Same as above.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/6f56b890138f16a67b5f61cb2238d75378a76497..22f65cbbbf36f1ae396038f457ca2266f5b52767#r72232076>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnlU1Px5M5rCEpafT5tCw-phmryocks5qZfGfgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak replied:
Sorry :innocent:
+1 on enabling, +1 on doing it in a follow-up PR

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+//    "comma-spacing": "error", // TODO enable?
+    "comma-style": ["error", "last"],
+    "eol-last": "error",
+//    "id-blacklist": ["error", "event"], // TODO enable?
```
I can live without this one :smiley:

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+    "no-unneeded-ternary": "error",
+//    "object-curly-spacing": ["error", "never"], // TODO enable?
+//    "object-property-newline": ["error", { "allowMultiplePropertiesPerLine": true }], // TODO enable?
+//    "operator-linebreak": ["error", "after", { "overrides": { "?": "before", ":": "before" }}], // TODO enable?
```
(Fwiw, I prefer no overrides :grin:)

mgol replied:
I added the overrides because I thought that was the current style (at
least I've seen a lot of them).

gkalpak replied:
I think there are many occurrences for `?`/`:`, but for other operators as well. I prefer being consistent throughout, so let's leave it for later (so we can bikeshed properly :stuck_out_tongue: )

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+//    "object-curly-spacing": ["error", "never"], // TODO enable?
+//    "object-property-newline": ["error", { "allowMultiplePropertiesPerLine": true }], // TODO enable?
+//    "operator-linebreak": ["error", "after", { "overrides": { "?": "before", ":": "before" }}], // TODO enable?
+//    "quotes": ["error", "single"], // TODO enable?
```
:tada:

Narretz replied:
Should be enabled imo

mgol replied:
>
> > +// "quotes": ["error", "single"], // TODO enable?
>
> Should be enabled imo
>
I'd like to enable it, too, but not in this PR. Note that we currently
don't have this rule enabled (and we have lots of violation) so I don't
feel it has to be a part of the migration.

gkalpak commented on .eslintrc.json:
```diff
@@ -0,0 +1,136 @@
+    "space-unary-ops": ["error", { "words": true, "nonwords": false }],
+    "unicode-bom": ["error", "never"]
+  }
+ }
```
Wrong indentation :stuck_out_tongue:

gkalpak commented on .jscsrc:
```diff
@@ -1,48 +0,0 @@
-{
-  "excludeFiles": ["src/ngLocale/**"],
```
I wonder why we excluded `ngLocale/**` and if we should keep doing it...

mgol replied:
>
> I wonder why we excluded ngLocale/** and if we should keep doing it...
>
Because they're largely auto-generated AFAIK. Most of them are JSONs,
though, so disabling a couple of rules fixes that. Note that only JSCS was
disabled, not JSHint so disabling checking these files in ESLint would
decrease our linting scope.

-- 
Michał Gołębiowski

gkalpak commented on benchmarks/largetable-bp/app.js:
```diff
@@ -64,13 +64,13 @@ app.controller('DataController', function($scope, $rootScope) {
 });
 
-var fn = function() { return 'x'};
+var fn = function() { return 'x';};
```
No space after `;`? :confused:

mgol replied:
Current rules don't cover this behavior so it was not auto-fixed.

gkalpak commented on src/ng/directive/select.js:
```diff
@@ -183,15 +185,12 @@ var SelectController =
     } else if (interpolateValueFn) {
       // The value attribute is interpolated
       optionAttrs.$observe('value', function valueAttributeObserveAction(newVal) {
-        var currentVal = self.readValue();
```
Did this have side-effects?

mgol replied:
>
> Did this have side-effects?
>
It shouldn't have, this method is just checking some stuff & returning a
value.

-- 
Michał Gołębiowski

gkalpak replied:
Yes, but I think we are overwriting the default implementation in `ngOptions` ([ngOptions.js#L506][1]).
I am not sure that there is indeed a usecase, where the side-effects in the new `readValue()` are necessary.

[1]: https://github.com/mgol/angular.js/blob/9bb014b44db3c46bf9a018d4d859fa63bb60a585/src/ng/directive/ngOptions.js#L506

mgol replied:
>
> Yes, but I think we are overwriting the default implementation in
> ngOptions (ngOptions.js#L506
> <https://github.com/mgol/angular.js/blob/9bb014b44db3c46bf9a018d4d859fa63bb60a585/src/ng/directive/ngOptions.js#L506>
> ).
>
> I am not sure that there is indeed a usecase, where the side-effects in
> the new readValue() are necessary.
>
Good point. I hope our unit tests would catch it. :-)

gkalpak commented on src/ngAnimate/.eslintrc.json:
```diff
@@ -1,7 +1,9 @@
+    "maxlen": "off", /* ngAnimate docs contain wide tables */
+    "new-cap": "off",
+  },
+
```
Is this inheriting the globals from `src/.eslintrc.json`?

mgol replied:
Yup, configs are merged up the tree up until "/" (or "C:\" or whatever you
have in Windows-land ;)) unless one of those configs has `"root": true` -
then the lookup ends on that file.

gkalpak commented on test/auto/injectorSpec.js:
```diff
@@ -256,36 +255,39 @@ describe('injector', function() {
 
         // Support: Chrome 50-51 only
-        // TODO (gkalpak): Remove when Chrome v52 is relased.
+        // TODO (gkalpak): Remove when Chrome v52 is released.
```
Does ESLint also catch spelling mistakes? :stuck_out_tongue:

mgol replied:
>
> Does ESLint also catch spelling mistakes? 😛
>
No, that's ESLint's padawan, also known as @mgol. :P

gkalpak commented on test/auto/injectorSpec.js:
```diff
@@ -1142,15 +1145,13 @@ describe('strict-di injector', function() {
   it('should always use provider as `this` when invoking a factory', function() {
     var called = false;
-    function factoryFn() {
+    function FactoryFn() {
```
Could we use `/* @this */` or something instead? It is weird to have a factory function capitalized :grin:

mgol replied:
Why? It is a constructor after all.

gkalpak replied:
In what sense? It is not `new`'d. It is `$injector.invoke`d.

mgol replied:
OK, I see what you mean.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -10006,9 +10007,7 @@ describe('$compile', function() {
-    it('should not sanitize attributes other than src', inject(function($compile, $rootScope) {
-      /* jshint scripturl:true */
-      element = $compile('<img title="{{testUrl}}"></img>')($rootScope);
+    it('should not sanitize attributes other than src', inject(function($compile, $rootScope) {      element = $compile('<img title="{{testUrl}}"></img>')($rootScope);
```
Missing newline.

gkalpak commented:
So far, this LGTM :+1: 

Except where otherwise noted, I like (eventually) enabling the rules marked as `TODO enable?`, but I think it is better to leave them disabled at first and discuss/enable them in follow-up PRs.

mgol committed:
chore(bower): change the name in bower.json

This avoids warnings with newer versions of Bower.

mgol committed:
chore(package.json): Remove the deprecated licenses field

The "licenses" field is deprecated in favor of the "license" field... which
we already have specified.

jbedard commented on src/auto/injector.js:
```diff
@@ -819,7 +818,8 @@ function createInjector(modulesToLoad, strictDi) {
           path.unshift(serviceName);
           cache[serviceName] = INSTANTIATING;
-          return cache[serviceName] = factory(serviceName, caller);
+          cache[serviceName] = factory(serviceName, caller);
```
~100 lines up (in `provider`) this was fixed by adding `(....)` around the expression instead. Why solve this one differently?

mgol replied:
I actually prefer this way most of the time, the intention is clearer and the JIT should optimize it anyway. I changed most of the things like I did here.

I think the parens "hack" is nice in case of regex matching where it's not so easy to change without introducing a repeated line of code.

mgol replied:
I've looked again and I probably did it differently above because the LHS expression was more complex so more would have to be repeated.

jbedard commented on src/ng/http.js:
```diff
@@ -909,7 +914,7 @@ function $HttpProvider() {
   var fetchBtn = element(by.id('fetchbtn'));
   var sampleGetBtn = element(by.id('samplegetbtn'));
-  var sampleJsonpBtn = element(by.id('samplejsonpbtn'));
+//  var sampleJsonpBtn = element(by.id('samplejsonpbtn'));
```
Delete? Or undo?

mgol replied:
It's unused because the test that was using it was commented out and ESLint was shouting about an unused variable.

gkalpak replied:
To avoid confusion, we could move this line inside the commented out test (where there is a comment with more context.

petebacondarwin commented on src/jqLite.js:
```diff
@@ -11,9 +11,10 @@
 
-/* global JQLitePrototype: true,
+/* global
+  JQLitePrototype: true,
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -11,9 +11,10 @@
+  JQLitePrototype: true,
   BOOLEAN_ATTR: true,
-  ALIASED_ATTR: true,
+  ALIASED_ATTR: true
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -135,9 +136,9 @@ JQLite._data = function(node) {
 
 
-var SPECIAL_CHARS_REGEXP = /([\:\-\_]+(.))/g;
+var SPECIAL_CHARS_REGEXP = /([:\-_]+(.))/g;
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -135,9 +136,9 @@ JQLite._data = function(node) {
+var SPECIAL_CHARS_REGEXP = /([:\-_]+(.))/g;
 var MOZ_HACK_REGEXP = /^moz([A-Z])/;
-var MOUSE_EVENT_MAP= { mouseleave: "mouseout", mouseenter: "mouseover"};
+var MOUSE_EVENT_MAP = { mouseleave: 'mouseout', mouseenter: 'mouseover' };
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -261,10 +262,9 @@ function jqLiteWrapNode(node, wrapper) {
-var jqLiteContains = window.Node.prototype.contains || function(arg) {
-  // jshint bitwise: false
+var jqLiteContains = window.Node.prototype.contains || /* @this */ function(arg) {
+  // eslint-disable-next-line no-bitwise
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -538,9 +538,8 @@ var JQLitePrototype = JQLite.prototype = {
       this.on('DOMContentLoaded', trigger); // works for modern browsers and IE9
       // we can not use jqLite since we are not done loading and jQuery could be loaded later.
-      // jshint -W064
+      // eslint-disable-next-line new-cap
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -645,7 +644,7 @@ forEach({
     if (BOOLEAN_ATTR[lowercasedName]) {
       if (isDefined(value)) {
-        if (!!value) {
+        if (value) {
```
LGTM

petebacondarwin commented on src/jqLite.js:
```diff
@@ -1069,6 +1068,7 @@ forEach({
 
 
 // Provider for private $$jqLite service
+/* @this */
```
LGTM

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -958,11 +958,12 @@ var _UNINITIALIZED_VALUE = new UNINITIALIZED_VALUE();
  * @description
  */
 $CompileProvider.$inject = ['$provide', '$$sanitizeUriProvider'];
+/* @this */
```
LGTM

petebacondarwin commented:
jqlite.js LGTM

petebacondarwin commented:
We should disable `no-void` rule. It is a valid use, IMO.

petebacondarwin commented:
`sanitize.js` LGTM

petebacondarwin commented:
`ngParseExt/module.js` LGTM

petebacondarwin commented on src/ngMessageFormat/messageFormatCommon.js:
```diff
@@ -34,7 +34,7 @@ function parseTextLiteral(text) {
     var unwatch = scope['$watch'](noop,
         function textLiteralWatcher() {
-          if (isFunction(listener)) { listener.call(null, text, text, scope); }
+          if (isFunction(listener)) { listener(text, text, scope); }
```
Not sure about this one.
I believe that the point is that we actively want to ensure that we do not leak any `this` value through to the user provided listener function.
I think we might even have a test in `ngParse` for this...

petebacondarwin commented on src/ngMessageFormat/messageFormatCommon.js:
```diff
@@ -51,14 +51,14 @@ function subtractOffset(expressionFn, offset) {
     unwatch = scope['$watch'](expressionFn,
         function pluralExpressionWatchListener(newValue, oldValue) {
-          if (isFunction(listener)) { listener.call(null, minusOffset(newValue), minusOffset(oldValue), scope); }
+          if (isFunction(listener)) { listener(minusOffset(newValue), minusOffset(oldValue), scope); }
```
Same as https://github.com/angular/angular.js/pull/14952/files#r72882347

petebacondarwin commented:
`sce.js` LGTM

petebacondarwin commented:
I have reviewed all the "sensitive" files. But not looked at everything.
The only outstanding issues, for me, are:
* disable the `no-void` rule and revert all the `void 0` -> `undefined` changes throughout the code
* revert the changes in `messageFormatCommon.js` around using `.call(null, listener)`

gkalpak commented on compare-master-to-stable.js:
```diff
@@ -9,72 +9,72 @@ var Q = require('q');
           results.push(res);
         });
-      }(args.pop()));
+      })(args.pop());
```
Does ESLint complain about the previous form?

mgol replied:
>
> Does ESLint complain about the previous form?
>
It depends on how you configure the wrap-iife rule:
http://eslint.org/docs/rules/wrap-iife.

I'm fine with changing it from `inside` to `outside` if that's the
preference.

gkalpak replied:
Aha, I didn't know you need to choose one or the other with ESLInt.
(I prefer `outside`, but I think it is a personal preference. `Inside` is fine.)

mgol replied:
You don't *have to* (there is the "any" option) but it's better to be
consistent.

On Mon, Aug 1, 2016 at 4:30 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In compare-master-to-stable.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72986746>:
>
> >            results.push(res);
> >          });
> > -      }(args.pop()));
> > +      })(args.pop());
>
> Aha, I didn't know you need to choose one or the other with ESLInt.
> (I prefer outside, but I think it is a personal preference. Inside is
> fine.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73#r72986746>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUngksIxUA4vy4b7kjKplGlXUwv6Fvks5qbgMDgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak commented on compare-master-to-stable.js:
```diff
@@ -85,44 +85,43 @@ var checkout = oneArg(exec('git checkout %s'));
-    filter(function (ver) {
-      return semver(ver).major == major;
+    filter(function(ver) {
+      return semver(ver).major === major;
```
Afaict, `major` will be a string and `semver(ver).major` will be a number, so `===` won't work.
I think it's fine to convert `major` to a number (since it is only used locally for comparison against `semver(ver).major`).

mgol replied:
Nice catch.

> I think it's fine to convert major to a number

Or we could stop splitting the string by ourselves and use the semver module for that since we're already using it in other places (like this line) for this very purpose.

mgol replied:
Fixed by using `semver` in line 105 as well. I haven't tested it but it's the same as in line 108 so I don't envision problems.

gkalpak commented on docs/app/assets/js/search-worker.js:
```diff
@@ -1,6 +1,7 @@
+'use strict';
+
+/* eslint-env worker */
+/* global importScripts, lunr */
```
Do we need these? Aren't we ignoring `docs/app/assets/**`?

gkalpak replied:
Actually, is there any reason why ignore assets?

mgol replied:
>
> Actually, is there any reason why ignore assets?
>
The reason is docs/app/assets/js/angular-bootstrap/** which contains code
pasted from AngularUI. I changed the ignore rules.

gkalpak commented on docs/app/src/examples.js:
```diff
@@ -1,8 +1,9 @@
 
 .directive('runnableExample', ['$templateCache', '$document', function($templateCache, $document) {
   var exampleClassNameSelector = '.runnable-example-file';
-  var doc = $document[0];
```
Actually we can (and should) remove the `$document` dependency altogether.
The `$templateCache` too.

mgol replied:
Done

gkalpak commented on docs/app/src/search.js:
```diff
@@ -23,43 +25,43 @@ angular.module('search', [])
         var totalAreas = 0;
-        for(var i in results) {
+        // eslint-disable-next-line no-unused-vars
+        for (var i in results) {
```
Wtf? Why not `var totalAreas = results.length`?

mgol replied:
>
> Wtf? Why not var totalAreas = results.length?
>
It's an object, not an array. I changed that locally to
`Object.keys(results).length`.

gkalpak commented on docs/content/guide/$location.ngdoc:
```diff
@@ -227,18 +227,18 @@ it('should show example', inject(
   },
   function($location) {
     // open http://example.com/base/index.html#!/a
-    $location.absUrl() == 'http://example.com/base/index.html#!/a'
```
This whole example/test is wrong anyway. But it's not related to this PR, so feel free to leave as is.

mgol replied:
Yeah, I'll leave it as it is here.

gkalpak commented on docs/content/guide/component.ngdoc:
```diff
@@ -388,7 +388,7 @@ Here is a tab pane example built from components:
   .component('myTabs', {
     transclude: true,
-    controller: function() {
+    controller: function Controller() {
```
Nitpick: Since this appears in the guide (which is supposed to promote good practices), I would name this `MyTabsController`.

mgol replied:
Changed

gkalpak commented on docs/content/guide/decorators.ngdoc:
```diff
@@ -356,6 +356,7 @@ those conditions.
               // create and store the getters using $parse
               var getters = exps.map(function(el) {
                 if (el) return $parse(el);
+                return undefined;
```
I think chagning the function body to `return el && $parse(el)` would be clearer for the reader.
Also, `el` seems quite unfitting. `exp` would be more intuitive (but this is unrelated to this PR :grin:)

mgol replied:
Changed

gkalpak commented on docs/content/guide/directive.ngdoc:
```diff
@@ -856,7 +856,7 @@ to which tab is active.
           transclude: true,
           scope: {},
-          controller: ['$scope', function($scope) {
+          controller: ['$scope', function Controller($scope) {
```
Nitpick: I would prefer `MyTabsController`.

mgol replied:
Fixed

gkalpak commented on docs/gulpfile.js:
```diff
@@ -120,35 +119,46 @@ gulp.task('doc-gen', ['bower'], function() {
 
-  var protractorConfig = getMergedJshintConfig('../docs/app/e2e/.jshintrc');
+  var protractorConfig = getMergedEslintConfig('../docs/app/e2e/.eslintrc.json');
+  protractorConfig.rules['no-unused-vars'] = ["error", {
```
Inconsistent quotes alert!!! :smiley:

mgol replied:
Fixed ;)

gkalpak commented on i18n/src/.eslintrc.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "env": {
+    "browser": false,
```
Do we need `"browser": false` is all `.eslintrc.json` files?

mgol replied:
>
> In i18n/src/.eslintrc.json
> <https://github.com/angular/angular.js/pull/14952#discussion_r72910207>:
>
> > @@ -0,0 +1,6 @@
> > +{
> > +  "env": {
> > +    "browser": false,
>
> Do we need "browser": false is all .eslintrc.json files?
>
Technically, no (and I'm afraid I'm not being consistent in this PR in this
regard); it's just ESLint by default extends rules up the tree and we have
code structure nested in funny ways so sometimes we have to overwrite
`browser: true` up the tree.

Maybe it'd be better to stop relying on the default inheritance and use
extends as we did with JSHint? Then we'd only have rules in the extended
files and the rest (env, globals) would be declared in all the files.

Thoughts? I can change the files like that if you think it'd be better.

gkalpak commented on i18n/src/closureI18nExtractor.js:
```diff
@@ -58,10 +60,12 @@ function extractCurrencySymbols(content) {
   for (var propName in goog.i18n) {
     var localeID = findLocaleId(propName, 'datetime');
     if (localeID) {
+      // eslint-disable-next-line no-unused-vars
```
Instead of this, you could change the `localeInfo[localeID].DATETIME_FORMATS...` to `info.DATETIME_FORMATS...`.

mgol replied:
>
> In i18n/src/closureI18nExtractor.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72910295>:
>
> >    eval(content.toString());
> >    for (var propName in goog.i18n) {
> >      var localeID = findLocaleId(propName, 'datetime');
> >      if (localeID) {
> > +      // eslint-disable-next-line no-unused-vars
>
> Instead of this, you could change the
> localeInfo[localeID].DATETIME_FORMATS... to info.DATETIME_FORMATS....
>
Nice catch!

gkalpak commented on i18n/src/closureSlurper.js:
```diff
@@ -3,14 +3,8 @@
 var Q  = require('q'),
     qfs  = require('q-io/fs'),
-    converter = require('./converter.js'),
-    util = require('./util.js'),
```
Actually, the whole `i18n/src/util.js` file could be removed. It's never used.

mgol replied:
This file was last touched 5 years ago... :D Removed.

mgol commented:
PR updated.

gkalpak commented on src/ng/controller.js:
```diff
@@ -142,8 +144,7 @@ function $ControllerProvider() {
 
-        var instantiate;
-        return instantiate = extend(function $controllerInit() {
+        var instantiate = extend(function $controllerInit() {
```
We don't need to even need the extra var. We could as well return directly without assigning.

gkalpak commented on src/ng/directive/ngModel.js:
```diff
@@ -1304,7 +1304,7 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
   return {
     restrict: 'A',
-    controller: ['$scope', '$attrs', function($scope, $attrs) {
+    controller: ['$scope', '$attrs', function Ctrl($scope, $attrs) {
```
Nitpick: Prefer `NgModelOptionsController` :innocent:

gkalpak commented on src/ng/directive/ngSwitch.js:
```diff
@@ -134,10 +134,10 @@ var ngSwitchDirective = ['$animate', '$compile', function($animate, $compile) {
      this.cases = {};
     }],
-    link: function(scope, element, attr, ngSwitchController) {
+    link: function(scope, element, attr, NgSwitchController) {
```
Is this necessary (here and below). We usually name these parameters using camelCase.

gkalpak commented on src/ng/parse.js:
```diff
@@ -1882,6 +1881,8 @@ function $ParseProvider() {
  /**
   * @ngdoc method
   * @name $parseProvider#setIdentifierFns
+  * @this
```
Ooc, why is this necessary? Doesn't ESLint allow methods of `@this` function to access `this`?

mgol replied:
>
> In src/ng/parse.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72947471>:
>
> > @@ -1882,6 +1881,8 @@ function $ParseProvider() {
> >   /**
> >    * @ngdoc method
> >    * @name $parseProvider#setIdentifierFns
> > +  * @this
>
> Ooc, why is this necessary? Doesn't ESLint allow methods of @this
> function to access this?
>
It's not necessary. :D Removed.

gkalpak commented on src/ng/q.js:
```diff
@@ -376,6 +378,7 @@ function qFactory(nextTick, exceptionHandler, errorOnUnhandledRejections) {
   }
 
   function processChecks() {
+    // eslint-disable-next-line no-unmodified-loop-condition
```
Is this necessary? As far as I understand, it shouldn't.

mgol replied:
>
> In src/ng/q.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72948617>:
>
> > @@ -376,6 +378,7 @@ function qFactory(nextTick, exceptionHandler, errorOnUnhandledRejections) {
> >    }
> >
> >    function processChecks() {
> > +    // eslint-disable-next-line no-unmodified-loop-condition
>
> Is this necessary? As far as I understand, it shouldn't.
>
I only added the comments to places where ESLint complained so they're all
necessary. Here the issue is that we're checking queueSize which is not
modified in the loop.

gkalpak replied:
Interesting. According to [this][1]:
> If a reference is inside of a binary expression or a ternary expression, this rule checks the result of the expression instead.

(And our expression should qualify as "binary expression".)

Nevermind, if it does complain, there's not much we can do :smile:

[1]: http://eslint.org/docs/rules/no-unmodified-loop-condition

gkalpak commented on src/ng/rootScope.js:
```diff
@@ -612,6 +618,7 @@ function $RootScopeProvider() {
               oldItem = oldValue[i];
               newItem = newValue[i];
 
+              // eslint-disable-next-line no-self-compare
```
We should have a helper for this thing :stuck_out_tongue:

gkalpak commented on src/ng/templateRequest.js:
```diff
@@ -86,7 +88,7 @@ function $TemplateRequestProvider() {
           transformResponse: transformResponse
         }, httpOptions))
-        ['finally'](function() {
+        .finally(function() {
```
This is intentional to make it possible for people to support some older browsers. Although the officially supported browsers aren't affected by this, people are able to support older browsers (e.g. older Androids) using polyfills and stuff like that, but `.finally`s were causing them problems. So, we had decided (iirc) to make their lives easier (even if we don't actively support such browsers), since it is a trivial change and doesn't affect anything else.

mgol replied:
>
> In src/ng/templateRequest.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72949259>:
>
> > @@ -86,7 +88,7 @@ function $TemplateRequestProvider() {
> >            cache: $templateCache,
> >            transformResponse: transformResponse
> >          }, httpOptions))
> > -        ['finally'](function() {
> > +        .finally(function() {
>
> This is intentional to make it possible for people to support some older
> browsers. Although the officially supported browsers aren't affected by
> this, people are able to support older browsers (e.g. older Androids) using
> polyfills and stuff like that, but .finallys were causing them problems.
> So, we had decided (iirc) to make their lives easier (even if we don't
> actively support such browsers), since it is a trivial change and doesn't
> affect anything else.
>
We already have lots of `.catch` occurrences in `src/` and Android 2.3
fails on that in exactly the same way. So it's already completely broken
there (parse errors === nothing works) at this very moment.

gkalpak replied:
Yeah, most/all of the them where introdiced as part of the Possibly Unhandled Rejection feature.
I don't think we have any in `v1.5.x` though. So, since we intent to backport this to 1.5.x, I wouldn't change this.

Narretz replied:
I agree, let's not change `[finally]` to `.finally` (same with catch). We can make a decision about "officially" stopping support for Android 2.3 in 1.6

mgol replied:
>
> In src/ng/templateRequest.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r73159282>:
>
> > @@ -86,7 +88,7 @@ function $TemplateRequestProvider() {
> >            cache: $templateCache,
> >            transformResponse: transformResponse
> >          }, httpOptions))
> > -        ['finally'](function() {
> > +        .finally(function() {
>
> I agree, let's not change [finally] to .finally (same with catch). We can
> make a decision about "officially" stopping support for Android 2.3 in 1.6
>
.catch has already been changed on master and this PR is against master.
I'm not opposed to requiring to quote keyword-like properties in 1.5 but
since on master it's already been done I'd rather do it as a followup in
the 1.5 version of this PR.

gkalpak commented on src/ngResource/resource.js:
```diff
@@ -425,8 +425,9 @@ function shallowClearAndCopy(src, dst) {
  */
 angular.module('ngResource', ['ng']).
-  provider('$resource', function() {
+  provider('$resource', function Resource() {
```
`ResourceProvider` would be more accurate :innocent:

gkalpak commented on src/ngSanitize/sanitize.js:
```diff
@@ -476,6 +477,7 @@ function $SanitizeProvider() {
           out(tag);
           out('>');
         }
+        // eslint-disable-next-line eqeqeq
```
I am pretty sure this one could be converted to `===`.

mgol replied:
>
> In src/ngSanitize/sanitize.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72962101>:
>
> > @@ -476,6 +477,7 @@ function $SanitizeProvider() {
> >            out(tag);
> >            out('>');
> >          }
> > +        // eslint-disable-next-line eqeqeq
>
> I am pretty sure this one could be converted to ===.
>
This is a sensitive file so I was afraid of touching it in this way. A next
security review would be necessary; if a few people say it's safe, I may
change it.

gkalpak commented on src/ngScenario/dsl.js:
```diff
@@ -468,7 +468,7 @@ angular.scenario.dsl('element', function() {
   });
 
-  return function(selector, label) {
+  return /* @this */ function(selector, label) {
```
We could as well disable `no-invalid-this` in this file.

mgol replied:
>
> In src/ngScenario/dsl.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72962680>:
>
> > @@ -468,7 +468,7 @@ angular.scenario.dsl('element', function() {
> >      };
> >    });
> >
> > -  return function(selector, label) {
> > +  return /* @this */ function(selector, label) {
>
> We could as well disable no-invalid-this in this file.
>
Good idea.

gkalpak commented on test/AngularSpec.js:
```diff
@@ -559,42 +559,42 @@ describe('angular', function() {
 
     it('should copy String() objects', function() {
-      /*jshint -W053 */
+      /* eslint-disable no-new-wrappers */
```
Ooc, wouldn't `esline-disable-next-line` work (and be cleaner)?

gkalpak commented on test/AngularSpec.js:
```diff
@@ -1412,7 +1409,7 @@ describe('angular', function() {
       var log = [];
       var self = {};
-      forEach(obj, function(val, key, collection) {
+      forEach(obj, /* @this */ function(val, key, collection) {
```
Is this necessry (considering `no-invalid-this` has been disabled)?

mgol replied:
It's a leftover from before its been disabled.

On Mon, Aug 1, 2016 at 1:37 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In test/AngularSpec.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72963406>:
>
> > @@ -1412,7 +1409,7 @@ describe('angular', function() {
> >        });
> >        var log = [];
> >        var self = {};
> > -      forEach(obj, function(val, key, collection) {
> > +      forEach(obj, /* @this */ function(val, key, collection) {
>
> Is this necessry (considering no-invalid-this has been disabled)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73#r72963406>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnk80ND-UFAJLgdIT7H0ifhVFV0uLks5qbdpvgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak commented on test/auto/injectorSpec.js:
```diff
@@ -469,15 +472,15 @@ describe('injector', function() {
         log += 'aConfig;';
       })
-      .provider('$a', function() {
+      .provider('$a', function Provider$a() {
```
Cheating :stuck_out_tongue:

gkalpak commented on test/helpers/testabilityPatch.js:
```diff
@@ -229,9 +230,9 @@ function sortedHtml(element, showNgClass) {
         for (var css in node.style) {
           var value = node.style[css];
-          if (isString(value) && isString(css) && css != 'cssText' && value && (1 * css != css)) {
+          if (isString(value) && isString(css) && css !== 'cssText' && value && (1 * css !== css)) {
```
The `!=` in `1 * css != css` was intentional.

gkalpak commented on test/ng/directive/ngRepeatSpec.js:
```diff
@@ -1141,7 +1143,7 @@ describe('ngRepeat', function() {
       expect($exceptionHandler.errors.shift().message).
           toMatch(
-            /^\[ngRepeat:dupes\] Duplicates in a repeater are not allowed\. Use 'track by' expression to specify unique keys\. Repeater: item in items, Duplicate key: object:3, Duplicate value: {}/);
+            /^\[ngRepeat:dupes] Duplicates in a repeater are not allowed\. Use 'track by' expression to specify unique keys\. Repeater: item in items, Duplicate key: object:3, Duplicate value: /);
```
Why remove the `{}` at the end (here and below)?

mgol replied:
>
> In test/ng/directive/ngRepeatSpec.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72969231>:
>
> > @@ -1141,7 +1143,7 @@ describe('ngRepeat', function() {
> >        scope.$digest();
> >        expect($exceptionHandler.errors.shift().message).
> >            toMatch(
> > -            /^\[ngRepeat:dupes\] Duplicates in a repeater are not allowed\. Use 'track by' expression to specify unique keys\. Repeater: item in items, Duplicate key: object:3, Duplicate value: {}/);
> > +            /^\[ngRepeat:dupes] Duplicates in a repeater are not allowed\. Use 'track by' expression to specify unique keys\. Repeater: item in items, Duplicate key: object:3, Duplicate value: /);
>
> Why remove the {} at the end (here and below)?
>
Because they weren't doing anything, it were just empty patterns which is
why ESLint was complaining about them.

mgol replied:
Ah, no, I'm wrong, they're just braces since there's nothing inside and it
was not ESLint but my IDE which was warning incorrectly. I'll restore them.

gkalpak commented on test/ng/parseSpec.js:
Wrong indentation (here and at the closing `})`).

gkalpak commented on test/ng/sceSpecs.js:
```diff
@@ -437,7 +437,7 @@ describe('SCE', function() {
         }).toThrowMinErr('$injector', 'modulerr', new RegExp(
              /Failed to instantiate module function ?\(\$sceDelegateProvider\) due to:\n/.source +
-             /[^[]*\[\$sce:iwcard\] Illegal sequence \*\*\* in string matcher\.  String: http:\/\/\*\*\*/.source));
+             /[^[]*\[\$sce:iwcard\] Illegal sequence \*\*\* in string matcher\. {2}String: http:\/\/\*\*\*/.source));
```
How come you didn't unescape `\]`? :smiley:

mgol replied:
It didn't shout at me so I haven't.

On Mon, Aug 1, 2016 at 2:41 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In test/ng/sceSpecs.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72970773>:
>
> > @@ -437,7 +437,7 @@ describe('SCE', function() {
> >            runTest({whiteList: ['http://***']}, null)();
> >          }).toThrowMinErr('$injector', 'modulerr', new RegExp(
> >               /Failed to instantiate module function ?\(\$sceDelegateProvider\) due to:\n/.source +
> > -             /[^[]*\[\$sce:iwcard\] Illegal sequence \*\*\* in string matcher\.  String: http:\/\/\*\*\*/.source));
> > +             /[^[]*\[\$sce:iwcard\] Illegal sequence \*\*\* in string matcher\. {2}String: http:\/\/\*\*\*/.source));
>
> How come you didn't unescape \]? 😃
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73#r72970773>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnkJmhhYGfHEGwaNC3ZMkUqCoKCnqks5qbemIgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -428,7 +428,7 @@ describe('ngMock', function() {
       it('should delegate exception to the $exceptionHandler service', inject(
           function($interval, $exceptionHandler) {
-        $interval(function() { throw "Test Error"; }, 1000);
+        $interval(function() { throw 'Test Error'; }, 1000);
```
No `new Error(...)` here?

mgol replied:
No, this rule is disabled in tests as it was too problematic and it's not
really that big an issue (it is in the compiled binary where you'd lose the
stack otherwise).

On Mon, Aug 1, 2016 at 2:46 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In test/ngMock/angular-mocksSpec.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72971268>:
>
> > @@ -428,7 +428,7 @@ describe('ngMock', function() {
> >
> >        it('should delegate exception to the $exceptionHandler service', inject(
> >            function($interval, $exceptionHandler) {
> > -        $interval(function() { throw "Test Error"; }, 1000);
> > +        $interval(function() { throw 'Test Error'; }, 1000);
>
> No new Error(...) here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73#r72971268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnsDAqR-3kI_LJUYjWRjSjREJcUpMks5qbeqPgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -949,7 +948,7 @@ describe('ngMock', function() {
             try {
               throw error;
-            } catch (e) {}
+            } catch (e) { /** do nothing **/}
```
No `/* empty */` here?

mgol replied:
Any comment suffices.

On Mon, Aug 1, 2016 at 2:46 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In test/ngMock/angular-mocksSpec.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72971346>:
>
> > @@ -949,7 +948,7 @@ describe('ngMock', function() {
> >            if (!error.stack) {
> >              try {
> >                throw error;
> > -            } catch (e) {}
> > +            } catch (e) { /** do nothing **/}
>
> No /* empty */ here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73#r72971346>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnmR8E7lsK2CgBZleNvCbAlM6nU9Iks5qbeqngaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak replied:
I know, but what happened to consistency :stuck_out_tongue:

gkalpak commented on i18n/spec/utilSpec.js:
```diff
@@ -1,16 +0,0 @@
-    expect(util.findLocaleId('DateTimeSymbols_en', 'datetime')).toBe('en');
-    expect(util.findLocaleId('NumberFormatSymbols_en_US', 'num')).toBe('en_US');
-  });
-});
```
Maybe these could be moved to `closureI18nExtractorSpec`. It has some tests for `findLocaleId`, but there are some more here.

mgol replied:
I removed the file as you wanted so these tests would no longer work.

On Mon, Aug 1, 2016 at 3:24 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In i18n/spec/utilSpec.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r72976317>:
>
> > -'use strict';
> > -
> > -var util = require('../src/util.js');
> > -
> > -describe('findLocaleId', function() {
> > -  it('should find localeId', function() {
> > -    expect(util.findLocaleId('', 'num')).toBeUndefined();
> > -    expect(util.findLocaleId('aa', 'datetime')).toBeUndefined();
> > -    expect(util.findLocaleId('aa', 'randomType')).toBeUndefined();
> > -    expect(util.findLocaleId('NumberFormatSymbols_en', 'datetime')).toBeUndefined();
> > -    expect(util.findLocaleId('DateTimeSymbols_en', 'num')).toBeUndefined();
> > -
> > -    expect(util.findLocaleId('DateTimeSymbols_en', 'datetime')).toBe('en');
> > -    expect(util.findLocaleId('NumberFormatSymbols_en_US', 'num')).toBe('en_US');
> > -  });
> > -});
>
> Maybe these could be moved to closureI18nExtractorSpec. It has some tests
> for findLocaleId, but there are some more here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73..197c8608ca2d44fec56ef25cc44c847ad7884627#r72976317>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnhWg0QtpL3BLzHc_dOETyqEZ6h5eks5qbfOngaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak replied:
`closureI18nExtractor` implements (and exports) a `findLocaleId` (which is just a slightly enhanced version of `util.findLocaleId`. There are some tests for `findLocaleId` in `closureI18nExtractorSpec`, but `util` had some more tests.
Anyway, this is minor - we didn't have the tests before and they are not related to this PR, so it's fine to leave things as they are now (we have piggybacked enough on this PR :stuck_out_tongue:)

mgol replied:
OK, done

On Tue, Aug 2, 2016 at 3:31 PM, Georgios Kalpakas <notifications@github.com>
wrote:

> In i18n/spec/utilSpec.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r73154184>:
>
> > -'use strict';
> > -
> > -var util = require('../src/util.js');
> > -
> > -describe('findLocaleId', function() {
> > -  it('should find localeId', function() {
> > -    expect(util.findLocaleId('', 'num')).toBeUndefined();
> > -    expect(util.findLocaleId('aa', 'datetime')).toBeUndefined();
> > -    expect(util.findLocaleId('aa', 'randomType')).toBeUndefined();
> > -    expect(util.findLocaleId('NumberFormatSymbols_en', 'datetime')).toBeUndefined();
> > -    expect(util.findLocaleId('DateTimeSymbols_en', 'num')).toBeUndefined();
> > -
> > -    expect(util.findLocaleId('DateTimeSymbols_en', 'datetime')).toBe('en');
> > -    expect(util.findLocaleId('NumberFormatSymbols_en_US', 'num')).toBe('en_US');
> > -  });
> > -});
>
> closureI18nExtractor implements (and exports) a findLocaleId (which is
> just a slightly enhanced version of util.findLocaleId. There are some
> tests for findLocaleId in closureI18nExtractorSpec, but util had some
> more tests.
> Anyway, this is minor - we didn't have the tests before and they are not
> related to this PR, so it's fine to leave things as they are now (we have
> piggybacked enough on this PR 😛)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/a35c7b1049edc8033c0e3a6eaeeda6cd34c99e73..197c8608ca2d44fec56ef25cc44c847ad7884627#r73154184>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnmnFOE5ikzqnQjJyNGgk_kWV5XUTks5qb0aZgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

gkalpak commented on .eslintignore:
```diff
@@ -1,7 +1,7 @@
 build/**
 docs/bower_components/**
-docs/app/assets/**
+docs/app/assets/js/angular-bootstrap/**
```
Similar change is needed in `Gruntfile.js`.

gkalpak commented on benchmarks/largetable-bp/app.js:
```diff
@@ -21,7 +23,9 @@ app.controller('DataController', function($scope, $rootScope) {
   $scope.numberOfBindings = totalRows * totalColumns * 2 + totalRows + 1;
   $scope.numberOfWatches = '?';
 
+  /* @this DataController */
```
The `DataController` part is not accurate.

gkalpak commented on test/helpers/privateMocks.js:
```diff
@@ -31,16 +31,13 @@ function xthey(msg, vals, spec) {
-    return version >= 10; //only IE10+ support keyframes / transitions
-  }
-  return true;
+  // Support: IE < 11
```
_11_  -->  _10_

gkalpak commented on test/e2e/tools/.eslintrc.json:
```diff
@@ -1,6 +1,8 @@
+
   "env": {
-    "node": true,
     "jasmine": false
```
Is this necessary?

gkalpak commented on test/AngularSpec.js:
```diff
@@ -1,7 +1,16 @@
 
+// Lots of typed array globals are used in this file and ESLint is
+// not smart enough to understand the `typeof !== 'undefined'` guards.
+/* eslint-env es6 */
```
This is a little too aggresive imo. Not terribly important though (since this is just a spec file).

gkalpak replied:
And it is considerably better that having it enabled for the whole `test/` directory :smiley:

gkalpak commented on docs/app/assets/js/angular-bootstrap/dropdown-toggle.js:
```diff
@@ -139,9 +139,9 @@ angular.module('ui.bootstrap.dropdown', [])
       // WAI-ARIA
       element.attr({ 'aria-haspopup': true, 'aria-expanded': false });
-      scope.$watch(dropdownCtrl.isOpen, function(isOpen) {
+      scope.$watch(dropdownCtrl.isOpen, function( isOpen ) {
```
Whaaat?!? :stuck_out_tongue:

gkalpak replied:
Oh, I see :disappointed:

gkalpak commented on .eslintrc-browser.json:
```diff
@@ -0,0 +1,17 @@
+  "env": {
+    // Note: don't set `"browser": true`; code in "src/" should be compatible with
+    // non-browser environments like Node.js with a custom window implementation
+    // like jsdom. All browser globals should be taken from window.
```
I would be happier if this didn't affect tests tbh :confused:

mgol replied:
>
> In .eslintrc-browser.json
> <https://github.com/angular/angular.js/pull/14952#discussion_r72982735>:
>
> > @@ -0,0 +1,17 @@
> > +{
> > +  "extends": "./.eslintrc-base.json",
> > +
> > +  "env": {
> > +    // Note: don't set `"browser": true`; code in "src/" should be compatible with
> > +    // non-browser environments like Node.js with a custom window implementation
> > +    // like jsdom. All browser globals should be taken from window.
>
> I would be happier if this didn't affect tests tbh 😕
>

My thinking was that we might not want to make it harder to run unit tests
in non-browser environments. It's mostly about taking all browser globals
from window, not directly. Is that so bad?

gkalpak replied:
No, it's not that bad. It's just an extra constraint that we have (and need to remember), which doesn't buy us much (we don't care about unit tests in non-browser environments).

But I am fine leaving it as is.

Narretz commented on src/ng/parse.js:
```diff
@@ -401,7 +400,7 @@ AST.prototype = {
   program: function() {
     var body = [];
-    while (true) {
+    for (;;) {
```
This is a bit unusual (at least to me) and could be explained with a comment. Or the no-constant-condition rule could be disabled for this line (while keeping while(true)

mgol replied:
I could perhaps disable the checkLoops option:
http://eslint.org/docs/rules/no-constant-condition#checkloops

On Fri, Aug 5, 2016 at 4:31 PM, Martin Staffa <notifications@github.com>
wrote:

> In src/ng/parse.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r73701186>:
>
> > @@ -401,7 +400,7 @@ AST.prototype = {
> >
> >    program: function() {
> >      var body = [];
> > -    while (true) {
> > +    for (;;) {
>
> This is a bit unusual (at least to me) and could be explained with a
> comment. Or the no-constant-condition rule could be disabled for this line
> (while keeping while(true)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/4ab2433bb71fe2bf692ca169c46363385a3806e4#r73701186>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnjmiORxRvtOdyVjgvr2NaYpP_954ks5qc0lbgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

Narretz replied:
Just for this line or globally?

mgol replied:
Globally. I don't like such exceptions, this code is not special so if
`while (true)` here is fine, I'd enable it everywhere.

I mostly cared about disallowing stuff like `if (false)`. :)

On Fri, Aug 5, 2016 at 4:40 PM, Martin Staffa <notifications@github.com>
wrote:

> In src/ng/parse.js
> <https://github.com/angular/angular.js/pull/14952#discussion_r73702677>:
>
> > @@ -401,7 +400,7 @@ AST.prototype = {
> >
> >    program: function() {
> >      var body = [];
> > -    while (true) {
> > +    for (;;) {
>
> Just for this line or globally?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular.js/pull/14952/files/4ab2433bb71fe2bf692ca169c46363385a3806e4#r73702677>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrUnhSjWj5tpkaIUXp2MgMPMdxczppyks5qc0tFgaJpZM4JU90D>
> .
>



-- 
Michał Gołębiowski

mgol replied:
OTH, I'd like to leave `while (false)` disallowed... Hmm.

On Fri, Aug 5, 2016 at 5:01 PM, Michał Gołębiowski <m.goleb@gmail.com>
wrote:

> Globally. I don't like such exceptions, this code is not special so if
> `while (true)` here is fine, I'd enable it everywhere.
>
> I mostly cared about disallowing stuff like `if (false)`. :)
>
> On Fri, Aug 5, 2016 at 4:40 PM, Martin Staffa <notifications@github.com>
> wrote:
>
>> In src/ng/parse.js
>> <https://github.com/angular/angular.js/pull/14952#discussion_r73702677>:
>>
>> > @@ -401,7 +400,7 @@ AST.prototype = {
>> >
>> >    program: function() {
>> >      var body = [];
>> > -    while (true) {
>> > +    for (;;) {
>>
>> Just for this line or globally?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/angular/angular.js/pull/14952/files/4ab2433bb71fe2bf692ca169c46363385a3806e4#r73702677>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABrUnhSjWj5tpkaIUXp2MgMPMdxczppyks5qc0tFgaJpZM4JU90D>
>> .
>>
>
>
>
> --
> Michał Gołębiowski
>



-- 
Michał Gołębiowski

mgol replied:
OK. `while (false)` is already covered by `no-unreachable`. PR updated.

On Fri, Aug 5, 2016 at 5:03 PM, Michał Gołębiowski <m.goleb@gmail.com>
wrote:

> OTH, I'd like to leave `while (false)` disallowed... Hmm.
>
> On Fri, Aug 5, 2016 at 5:01 PM, Michał Gołębiowski <m.goleb@gmail.com>
> wrote:
>
>> Globally. I don't like such exceptions, this code is not special so if
>> `while (true)` here is fine, I'd enable it everywhere.
>>
>> I mostly cared about disallowing stuff like `if (false)`. :)
>>
>> On Fri, Aug 5, 2016 at 4:40 PM, Martin Staffa <notifications@github.com>
>> wrote:
>>
>>> In src/ng/parse.js
>>> <https://github.com/angular/angular.js/pull/14952#discussion_r73702677>:
>>>
>>> > @@ -401,7 +400,7 @@ AST.prototype = {
>>> >
>>> >    program: function() {
>>> >      var body = [];
>>> > -    while (true) {
>>> > +    for (;;) {
>>>
>>> Just for this line or globally?
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/angular/angular.js/pull/14952/files/4ab2433bb71fe2bf692ca169c46363385a3806e4#r73702677>,
>>> or mute the thread
>>> <https://github.com/notifications/unsubscribe-auth/ABrUnhSjWj5tpkaIUXp2MgMPMdxczppyks5qc0tFgaJpZM4JU90D>
>>> .
>>>
>>
>>
>>
>> --
>> Michał Gołębiowski
>>
>
>
>
> --
> Michał Gołębiowski
>



-- 
Michał Gołębiowski

Narretz commented:
I've looked at parse.js, jqlite.js, sanitize.js and compile.js, and they LGTM except a tiny question in parse.js

gkalpak commented:
@Narretz, what about `sce.js`?

Narretz commented:
And sce.js :>
LGTM now with the parse.js change

gkalpak commented:
Still LGTM2 :smile:

mgol committed:
chore(package.json): upgrade Protractor

The previous version depended on a vulnerable request version.

Ref gh-14961

mgol committed:
chore(package.json): update Karma & BrowserStack-related packages

mgol commented:
All commits except the ESLint one backported to `v1.5.x`. The last one is handled in PR #14993; please review.
